sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/CSVInferSchema.scala

```diff
@@ -108,14 +109,38 @@ private[csv] object CSVInferSchema {
   }
 
   private def tryParseDouble(field: String): DataType = {
-    if ((allCatch opt field.toDouble).isDefined) {
+    val doubleTry = allCatch opt field.toDouble
```

Comment:
Ah.. numeric types with fractions can be also `Decimal`. It has precision and scale.

```scala
import java.math.BigDecimal
scala> BigDecimal.valueOf(1.1111)
res4: java.math.BigDecimal = 1.1111

scala> BigDecimal.valueOf(1.1111).precision
res6: Int = 5

scala> BigDecimal.valueOf(1.1111).scale
res7: Int = 4
```

```scala
import java.math.BigDecimal
scala> BigDecimal.valueOf(1)
res5: java.math.BigDecimal = 1

scala> BigDecimal.valueOf(1).precision
res8: Int = 1

scala> BigDecimal.valueOf(1).scale
res9: Int = 0
```

`DoubleType` with fractions can lose precision if it has too many.